### PR TITLE
feat: register rules against specific formats

### PR DESCRIPTION
**Pull request by P0lip**
Closes https://github.com/stoplightio/spectral/issues/369

**Checklist**

- [x] Tests added / updated
- [ ] Docs added / updated

**Does this PR introduce a breaking change?**

- [ ] Yes
- [x] No

Heads up - I haven't written any documentation yet, but if we all agree on the proposed API, I'm more than happy to give it a stab.

A few words regarding the implementation I decided to go with.
The PR features support for both of the ideas described in the linked GH issue:
- rule-scoped formats as suggested by @philsturgeon 
- ruleset-scoped formats as suggested by @P0lip

There hasn't been anything said in terms of inheritance, so I decide to keep it simple and make formats ruleset-specific, with no inheritance at all.

Moreover, I hesitated from registering any default formats such as OAS2 or OAS3 globally and made them specific to CLI version only.
I exposed 2 simple functions that can be used for the spec detection, so if one wants to register oas2/oas3 formats, it should be pretty straightforward.

### CODE REVIEW

File: src/cli/commands/lint.ts [hunk 1]
```diff
@@ -11,6 +11,7 @@ import { json, stylish } from '../../formatters';
 import { readParsable } from '../../fs/reader';
 import { httpAndFileResolver } from '../../resolvers/http-and-file';
 import { getDefaultRulesetFile } from '../../rulesets/loader';
+import { isOAS2Spec, isOAS3Spec } from '../../rulesets/lookups';
```
Comment (philsturgeon) [inline src/cli/commands/lint.ts]: Some terminology style guide stuff here, instead of saying "Spec" (JSON is a spec, YAML is a spec, HTTP is a spec, we dont say spec after all of them), and the S in OAS is already "spec" anyway.

API Descriptions are documents written in special formats, so there is a description Document - the file(s) - and description Formats (OpenAPI, RAML, etc). 

Can we try and make sure that at least our new code is consistent with this style guide, and maybe boy scout anything internal we bump into on the way?

tldr: `isOpenApiV2` & `isOpenApiV3`

Comment (philsturgeon) [reply src/cli/commands/lint.ts]: oh we should probably make them isOpenApiv2_0, isOpenApiv3_0, isOpenApiv3_1, etc or something less ugly. People will want to have special rules for 3.0 like "hey upgrade to 3.1" lol

Comment (P0lip) [reply src/cli/commands/lint.ts]: Good catch. Didn't pay much attention to that, but you are absolutely right about it.
Will do.

**Comment by philsturgeon** [comment]
Btw the end goal here is to let people extend `spectral: oas` and it’ll all just magically work. I think we should merge all the rules into a single ruleset, and leave oas2 and oas3 there, empty, both just extending that main one. It’ll all work the same because yay extension.

File: src/cli/commands/lint.ts [hunk 2]
```diff
@@ -150,6 +151,24 @@ async function lint(name: string, flags: ILintConfig, command: Lint, rules?: Rul
     mergeKeys: true,
   });
   const spectral = new Spectral({ resolver: httpAndFileResolver });
+  spectral.registerFormat('oas-2', document => {
```
Comment (philsturgeon) [inline src/cli/commands/lint.ts]:
```suggestion
  spectral.registerFormat('oas2', document => {
```

File: src/cli/commands/lint.ts [hunk 3]
```diff
@@ -150,6 +151,24 @@ async function lint(name: string, flags: ILintConfig, command: Lint, rules?: Rul
...
+    if (isOpenApiv2_0(document)) {
+      command.log('OpenAPI 2.0 (Swagger) detected');
+      return true;
+    }
+
+    return false;
+  });
+
+  spectral.registerFormat('oas-3', document => {
```
Comment (philsturgeon) [inline src/cli/commands/lint.ts]:
```suggestion
  spectral.registerFormat('oas-3.0, document => {
```

File: src/cli/commands/lint.ts [hunk 4]
```diff
@@ -150,6 +151,24 @@ async function lint(name: string, flags: ILintConfig, command: Lint, rules?: Rul
...
+    if (isOpenApiv3(document)) {
```
Comment (philsturgeon) [inline src/cli/commands/lint.ts]: It’d more important for v3 to have the _0 because it’s gonna get v3.1 at some point. Should we leave them as OpenApiv2 and OpenApiV3 then just add _1 when it comes out? If so we can probably ignore the above comments on format names.

Comment (P0lip) [reply src/cli/commands/lint.ts]: Having `oas2`, `oas3` and potentially `oas3.1` seems like a reasonable choice.
Are we sticking to this?

**Comment by philsturgeon** [comment]
Let’s do that :)

--
Phil Sturgeon
@philsturgeon

**Comment by P0lip** [comment]
@philsturgeon 
Thank you for taking care of docs!
